devices/src/virtio/mmio.rs

```diff
@@ -33,20 +33,60 @@ const MMIO_VERSION: u32 = 2;
 /// Optionally, a virtio device can implement device reset in which it returns said resources and
 /// resets its internal.
 pub trait VirtioDevice: Send {
+    /// Get the available features offered by device.
+    fn avail_features(&self) -> u64;
+
+    /// Get acknowledged features of the driver.
+    fn acked_features(&self) -> u64;
+
+    /// Set acknowledged features of the driver.
+    /// This function must maintain the following invariant:
+    /// - self.get_avail_features() & self.get_acked_features() = self.get_acked_features()
+    fn set_acked_features(&mut self, acked_features: u64);
+
     /// The virtio device type.
     fn device_type(&self) -> u32;
 
     /// The maximum size of each queue that this device supports.
     fn queue_max_sizes(&self) -> &[u16];
 
     /// The set of feature bits shifted by `page * 32`.
-    fn features(&self, page: u32) -> u32 {
-        let _ = page;
-        0
+    fn paged_features(&self, page: u32) -> u32 {
+        let avail_features = self.avail_features();
+        match page {
+            // Get the lower 32-bits of the features bitfield.
+            0 => avail_features as u32,
+            // Get the upper 32-bits of the features bitfield.
+            1 => (avail_features >> 32) as u32,
+            _ => {
+                warn!("Received request for unknown features page.");
+                0u32
+            }
+        }
     }
 
     /// Acknowledges that this set of features should be enabled.
-    fn ack_features(&mut self, page: u32, value: u32);
+    fn paged_acked_features(&mut self, page: u32, value: u32) {
```

Comment:
I do not like the `paged` suffix from the functions name. The page argument dictates which bits should be taken into account (high or low). So it plays the role of a select.
I think the names of the functions should go like this:
`fn avail_features` -> gets available features, this is a getter so it should have the name of the field it gets
`fn acked_features` -> Ok
`fn select_acked_features` -> function for selecting acknowledged features using a select value
`fn select_avail_features` -> get available features using a select value

Reply:
`select_*` sounds like the action of selecting `*`.

The virtio documentation refers the upper and lower 32bit parts of features as `pages`. How about:

`fn avail_features() -> u64` - gets `self.avail_features`
`fn set_avail_features(val: u64)` - sets `self.avail_features`
`fn acked_features() -> u64` - gets already acked features (`self.acked_features`)
`fn set_acked_features(val: u64)` - sets already acked features (`self.acked_features`)

`fn ack_features_by_page(page: u32, val: u32)` - the action of acknowledging the features in `val` from page `page`
`fn avail_features_by_page(page:u32)` - gets avail features in page `page`

?

Notice the difference in the ACK case, getter and setter work on **acked** values, while the `fn ack_features_by_page()` has **ack** as a verb, the function is doing the action of acknowledging.

Reply:
I am inclined to @acatangiu suggestion. Also, the `fn ack_features_by_page()` method suggests verbatim the action which it should be responsible of, information which was lost once I've changed the original method name.

Reply:
I am ok with that proposal too.

Reply:
Updated as discussed.